pkg/services/ngalert/image/service.go

```diff
@@ -0,0 +1,118 @@
+	ErrNoPanel = errors.New("no panel")
+)
+
+const (
+	screenshotTimeout  = 10 * time.Second
+	screenshotCacheTTL = 15 * time.Second
+)
+
+// ScreenshotImageService takes screenshots of the panel for an alert rule and
+// saves the image in the store. The image contains a unique token that can be
+// passed as an annotation or label to the Alertmanager.
+type ScreenshotImageService struct {
+	screenshots screenshot.ScreenshotService
+	store       store.ImageStore
+}
+
+func NewScreenshotImageService(screenshots screenshot.ScreenshotService, store store.ImageStore) ImageService {
+	return &ScreenshotImageService{
+		screenshots: screenshots,
+		store:       store,
+	}
+}
+
+// NewScreenshotImageServiceFromCfg returns a new ScreenshotImageService
+// from the configuration.
+func NewScreenshotImageServiceFromCfg(cfg *setting.Cfg, metrics prometheus.Registerer,
+	db *store.DBstore, ds dashboards.DashboardService, rs rendering.Service) (ImageService, error) {
+	if !cfg.UnifiedAlerting.Screenshots.Enabled {
+		return &ScreenshotImageService{
+			screenshots: &screenshot.ScreenshotUnavailableService{},
+		}, nil
+	}
+
+	s := screenshot.NewBrowserScreenshotService(ds, rs)
+	if cfg.UnifiedAlerting.Screenshots.UploadExternalImageStorage {
+		u, err := imguploader.NewImageUploader()
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize uploading screenshot service: %w", err)
+		}
+		s = screenshot.NewUploadingScreenshotService(metrics, s, u)
+	}
+	s = screenshot.NewRateLimitScreenshotService(s, cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots)
+	s = screenshot.NewSingleFlightScreenshotService(s)
+	s = screenshot.NewCachableScreenshotService(metrics, screenshotCacheTTL, s)
+	s = screenshot.NewObservableScreenshotService(metrics, s)
+
+	return &ScreenshotImageService{
+		store:       db,
+		screenshots: s,
+	}, nil
+}
+
+// NewImage returns a screenshot of the panel for the alert rule. It returns
+// ErrNoDashboard if the alert rule does not have a dashboard and ErrNoPanel
+// when the alert rule does not have a panel in a dashboard.
+func (s *ScreenshotImageService) NewImage(ctx context.Context, r *ngmodels.AlertRule) (*store.Image, error) {
+	if r.DashboardUID == nil {
+		return nil, ErrNoDashboard
+	}
+	if r.PanelID == nil || *r.PanelID == 0 {
+		return nil, ErrNoPanel
+	}
+
+	screenshot, err := s.screenshots.Take(ctx, screenshot.ScreenshotOptions{
+		Timeout:      screenshotTimeout,
+		DashboardUID: *r.DashboardUID,
+		PanelID:      *r.PanelID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to take screenshot: %w", err)
+	}
+
+	v := store.Image{
+		Path: screenshot.Path,
+		URL:  screenshot.URL,
+	}
+	if err := s.store.SaveImage(ctx, &v); err != nil {
+		return nil, fmt.Errorf("failed to save image: %w", err)
+	}
+
+	return &v, nil
+}
+
+type NoopImageService struct{}
+
+func (s *NoopImageService) NewImage(ctx context.Context, r *ngmodels.AlertRule) (*store.Image, error) {
+	return &store.Image{}, nil
+}
```

Comment:
I wonder why it returns an empty struct instead of a nil.

Reply:
I'm up for deleting this service, personally. It caused me some testing headaches because it would fill empty structs in messages for any request. That's not a "Noop". It's why I ended up writing that `UnavailableImageService` later on. If I needed something for tests I'd rather just write a fake that stashes structs on save and does iterations for search.